### PR TITLE
Fixed and redefined total peering time (%<tt)

### PR DESCRIPTION
**Pull request by rousskov**
This change adjusts `%[http::]<tt` definition to include peer selection
stage (with any relevant DNS lookups). That stage runs in parallel with
connection establishment attempts, so excluding portions of that stage
made interpreting logged delays more difficult (and code more complex).

Also documented how cache refresh (TCP_REFRESH_MODIFIED) affects `%<tt`.
Fetching a new version of the stale object creates a second FwdState
object, effectively repeating peer selection activities and request
forwarding attempts. `%<tt` includes the time spent by those activities.

The following three bugs were fixed (at least):

* Squid logged dash `%<tt` values for pinned connections (e.g.,
  connections bumped at SslBump step 2 or 3) because Squid did not call
  HierarchyLogEntry::startPeerClock() for pinned connections.

* Logged `%<tt` values for connections spliced at SslBump step 2 or 3
  included TCP connection establishment time and TLS Server Hello
  fetching time (if any) but were missing the rest of Squid-peer
  communication (i.e. when Squid was tunneling bytes to and from the
  peer) because Squid stopped counting after calling switchToTunnel().

* Squid logged dash `%<tt` values for external_acl_type and most other
  directives that support logformat codes because ALE::hier data member
  is unset until access logging, when prepareLogWithRequestDetails()
  copies HttpRequest::hier to ALE. This is a short-term fix. A proper
  one deserves a dedicated change (that should also address similar
  concerns for other HierarchyLogEntry-derived %codes).

### CODE REVIEW

File: src/cf.data.pre
Comment (rousskov) [inline src/cf.data.pre:5024]: This PR preserves Squid behavior documented in this paragraph. The cumulative nature of `%<tt` stats across these cache refreshing attempts was not documented.

File: src/cf.data.pre
Comment (rousskov) [inline src/cf.data.pre:5011]: A secondary activity may start before the transaction peering timer when a transaction is joining an already pending secondary activity. For example, an HTTP request handled by FwdState (`%<tt` timer) may join an already ongoing DNS lookup for the requested domain name (`%dt` timer).

Similarly, a secondary activity may continue beyond a peering transaction phase. For example, a DNS A lookup may still be pending while the transaction proceeds with the DNS AAAA lookup result received earlier (due to Happy Eyeballs optimizations).

These natural imperfect/unpredictable concurrent activity/transaction overlaps imply that `%<tt - %dt` math does not always work as more "primitive" sequential processing models may suggest. In some cases, it may even yield negative results. We cannot really "fix" that, but we can make `%<tt` much simpler by not trying to exclude the "initial" DNS lookups from it (like the official code does, leading to admin confusion and Squid bugs). This PR adjusts `%<tt` definition and code accordingly.

**Review by rousskov — COMMENTED** [review]
This work was triggered by admin complaints about missing `%<tt` values _and_ `%<tt` values that cannot be reliably used for basic "proxy induced delay" calculations (math yielding negative values and such). The (essentially minor) adjustment of the `%<tt` definition allowed to address some of those problems, even though calculating "proxy induced delay" remains a controversial and contested area, with many reasonable solutions and no solution that can cover all corner cases.

The adjustment also allowed to greatly simplify `%<tt` statistics collection, not just fixing old bugs, but making code more reliable across future refactoring changes.

This review annotates this PR but does not request any changes.

File: src/tunnel.cc
Comment (eduard-bagdasaryan) [inline src/tunnel.cc:410]: We could simplify a little (eliminating referencing/dereferencing) by making guaranteedRequest() return a pointer: its very name states that there is an existing request.

Comment (rousskov) [reply src/tunnel.cc:410]: Yes, we could, but I did not want a function that guarantees request existence to return a pointer to that existing request: If the object is guaranteed to exist, its accessor should return a reference rather than a pointer. Either way, this code is problematic as detailed in https://github.com/squid-cache/squid/pull/1828#discussion_r1623108720. I hope it will go away with changes discussed in that change request.

Said that, if others insist, I will refactor the helper function to remove this `&`.

**Comment by rousskov** [comment]
> Removing the merge tag since this has not cleared the merge procedure criteria (10 days or 1+ non-author votes)

Re-adding M-cleared-for-merge label. It does not make sense to remove that label for the reason specified above because the label does not violate (and is _independent_ from) other "procedure criteria", including various timeouts. Removing the label (for that reason) has negative side effects: It forces folks to needlessly remember to add it later, wasting human time and increasing PR merging delays.

If you need to review a PR, request your own review. If you want another person to review a PR, request their review. Any request for review blocks the PR from merging (and one may remove the M-cleared-for-merge label in that case because it does little good when the PR is waiting for review(s) -- the last reviewer can clear the PR for merging as needed). In other cases, a PR with a positive vote (implicit or explicit) and without negative ones, may have an M-cleared-for-merge label while it is waiting for various timeouts to expire (or other significant PR events).

If you do not want to merge your own PR until somebody else (anybody) reviews it, do not clear it for merge, add a S-waiting-for-more-reviewers label, and a comment explaining this (rare) situation. That is what that label is for -- attract other reviewers to a PR that (effectively) does not have enough votes to be merged.

**Review by eduard-bagdasaryan — APPROVED** [review]
LGTM.

**Comment by yadij** [comment]
> > Removing the merge tag since this has not cleared the merge procedure criteria (10 days or 1+ non-author votes)
> 
> Re-adding M-cleared-for-merge label. It does not make sense to remove that label for the reason specified above because the label does not violate (and is _independent_ from) other "procedure criteria", including various timeouts. 

Strawman argument there. Other criteria are irrelevant.

Merge approval requires one of:
 A) 3 total votes (anyone)
 B) 1 non-author core vote
 C) 10 days since last code update to PR

This PR has (still today) not yet achieved any of those. Marking it for automation to take over (ie no more human input required) is wrong.

> Removing the label (for that reason) has negative side effects: It forces folks to needlessly remember to add it later, wasting human time and increasing PR merging delays.

 ... OR, one could follow the github discussion notifications.

> 
> If you need to review a PR, request your own review. If you want another person to review a PR, request their review.

I did request more review ... by setting that "waiting-for-more-reviewers" label which you cleared. It should have more eyes-on but a) anyone will do, and b) I might not have time, and c) not strongly enough to block the merge more than the 10+ day limit imposes.

To put it another way;  what we have going on in this PR with the two of you working on that long set of changes (and Factory internal reviews?) is essentially the same as one author writing code, submitting a PR and marking it cleared for merge immediately without review. 
 While I do trust both of you not to be malicious, it is still good/best practice to have **someone** unfamiliar with the changes to eyeball it before approval is granted.


> Any request for review blocks the PR from merging (and one may remove the M-cleared-for-merge label in that case because it does little good when the PR is waiting for review(s) -- the last reviewer can clear the PR for merging as needed). 

IIRC, that is a new take we have not discussed.

AIUI, we allow anyone to set the label IF (and only IF) the procedure requirements for setting it are met. Most of the time it is one of us in the core team doing the review - which meets that (B) criteria immedately passes and we can set it on a pass vote


> In other cases, a PR with a positive vote (implicit or explicit) and without negative ones, may have an M-cleared-for-merge label while it is waiting for various timeouts to expire (or other significant PR events).

That is abusing Anubis and should ideally not be happening. Reviewers should at least wait until the automated tests have passed before spending time on review. A portability failure may cause major changes to the logic/design and re-reviewing such things is becoming a noticeable waste of our time.
 [ yes I am guilty of doing this early-review, and pay the price in wasted time ]

> 
> If you do not want to merge your own PR until somebody else (anybody) reviews it, do not clear it for merge, add a S-waiting-for-more-reviewers label, and a comment explaining this (rare) situation. That is what that label is for -- attract other reviewers to a PR that (effectively) does not have enough votes to be merged.

Rule of thumb is that authors should not be clearing their own PRs, reviewers should be doing that. Though **after** the 10-day criteria passes it is reasonable to see that happen.

File: src/format/Format.cc
Comment (yadij) [reply src/format/Format.cc:647]: IMO code should not be doing this conversion at all. In here at least it should be displaying the duration object, not a timeval.

**Review by yadij — APPROVED** [review]

**Review by yadij — DISMISSED** [review]
sorry, misclick on the above.

**Comment by rousskov** [comment]
> Merge approval requires one of:
> A) 3 total votes (anyone)
> B) 1 non-author core vote
> C) 10 days since last code update to PR

I will not go into details of the _actual_ criteria in my response; the ABC version above differs from the [official](https://wiki.squid-cache.org/MergeProcedure#votes) [one](https://wiki.squid-cache.org/MergeProcedure#votes) in many ways, but I do not want this discussion to digress into those completely unrelated details (that are important in other contexts, but not in clearing M-cleared-for-merge misunderstandings).


> This PR has (still today) not yet achieved any of those. Marking it for automation to take over (ie no more human input required) is wrong.

There is absolutely nothing wrong with merging a PR approved by at least one core developer after a 10-day timeout without any further human input. Which is exactly what would have happened here if all core developers disappeared right after this PR was approved by me.

Very roughly speaking, GitHub+Anubis merging algorithm is this:

```C++
if (A and B and C and ... Y and Z) {
    merge PR
}
```

M-cleared-for-merge label is just one of the many conditions that GitHub and Anubis check before merging a PR. They include the conditions derived from A), B), and C) approval conditions you have listed above (e.g., 10-day timeout) as well as conditions like "all required tests passed" or "PR title is short enough". Adding M-cleared-for-merge label satisfies exactly one of those A..Z conditions. Nothing more! It does not overwrite any voting requirements or accelerate any timeouts. None of that!

Side note: I have probably mentioned that before somewhere, but reacting to M-cleared-for-merge is an optional Anubis feature. Anubis will work correctly -- obeying all Squid Project PR merging requirements it is told to obey -- with that feature turned off. There are large, busy open source projects where merge bots do not wait for any such label to be set; they merge immediately when all labeling-unrelated conditions (e.g., voting and testing) are satisfied. We enabled this Anubis feature just as an extra layer of protection or safety mechanism to reduce chances of _accidental_ PR mergers (because Squid merge process involves timeouts and other conditions that most humans can occasionally misjudge and because humans occasionally misclick).


> > Removing the label (for that reason) has negative side effects: It forces folks to needlessly remember to add it later, wasting human time and increasing PR merging delays.
 
> ... OR, one could follow the github discussion notifications.

In the scope of my assertion about negative side effects:

* Following notifications is not an option today: There are simply no notifications for an expired timeout.

* Following notifications is not a good goal for future work: We do not want to create an interface where a human action is always required after a notification about an expired timeout pops up because such an interface will always create completely unnecessary human work and merging delays.

 
> > If you need to review a PR, request your own review. If you want another person to review a PR, request their review.

> I did request more review ... by setting that "waiting-for-more-reviewers" label

S-waiting-for-more-reviewers label is not related to my rule of thumb quoted above. That rule is about a reviewer that needs to review a PR (but cannot do it now) or a person that wants another (specific) person to review a PR. It is _not_ about attracting attention of other potential reviewers.

All open PRs invite reviews by default (no special labeling is required or wanted!), so I recommend adding a brief GitHub comment when adding S-waiting-for-more-reviewers label to clarify what you expect to happen (i.e. why and how we should deviate from that natural default). In most cases, that label should not be added to PRs without negative votes that are approved by a core developer. And in exceptional cases, an explanation is warranted.


> [This PR] should have more eyes-on but a) anyone will do, and b) I might not have time, and c) not strongly enough to block the merge more than the 10+ day limit imposes.

(a) is true for pretty much any PR (no special labeling is needed), (b) is not reflected well by S-waiting-for-more-reviewers -- the label is meant to encourage reviews by _others_, and (c) is not going to work by labeling PRs: There is currently no mechanism to block a PR for 10 days. If you want to propose new mechanisms to better address that use case, let's discuss, but please do not remove M-cleared-for-merge labels from unblocked PRs approved by another core developer.

 
> To put it another way; what we have going on in this PR with the two of you working on that long set of changes (and Factory internal reviews?) is essentially the same as one author writing code, submitting a PR and marking it cleared for merge immediately without review.

Correction: Not "without review" but "without independent review" (or some such). PR code _was_ reviewed by two developers, one of which is a core developer.

As for the rest of it, yes, one can view the situation that way, but the PR still has to obey the 10 day timeout, like any other PR with just one core developer approval. There is nothing exceptional going on here! M-cleared-for-merge label simply tells other core developers (and Anubis) that this PR has been checked by a core developer (i.e. the human assigning that label) and will be auto-merged _if_ it satisfies all the usual criteria, including timeouts. Nothing more.


> While I do trust both of you not to be malicious, it is still good/best practice to have **someone** unfamiliar with the changes to eyeball it before approval is granted.

Absolutely! That is (one of the reasons) why we have a 10 day timeout. That timeout balances the need for a second pair of core developer eyes with the need to make progress when core developers cannot be compelled to post prompt reviews. M-cleared-for-merge has no effect on that (or any other) timeout.

 
> > Any request for review blocks the PR from merging (and one may remove the M-cleared-for-merge label in that case because it does little good when the PR is waiting for review(s) -- the last reviewer can clear the PR for merging as needed).

> IIRC, that is a new take we have not discussed.

What is a "new take", that review requests block PRs? That condition was in place since 2018 when the very first version of Anubis was deployed. Anubis even adds "waiting for requested reviews" PR Approval status to GitHub checks in that case (if all other conditions are satisfied). The impact of review requests is also explicitly disclosed in Anubis [documentation](https://github.com/measurement-factory/anubis/blob/master/README.md#voting-and-pr-approvals), of course.

It is difficult for me to imagine a reasonable scheme that would _not_ block a PR when there is a pending core developer review request. AFAICT, such an implementation would require us to vote negatively just to say "please wait for my review". Not a good UX for a very common PR state!

 
> AIUI, we allow anyone to set the label IF (and only IF) the procedure requirements for setting it are met.

AFAIK, only core developers (and GitHub accounts with similar privileges) can set PR labels.

M-cleared-for-merge label should be set when the person setting it believes this PR can be merged after satisfying all merge preconditions (voting, timeouts, tests, etc.). It is just an extra safety check.

The label is also occasionally useful when, for example, a core developer wants to approve a PR (to explicitly record their vote and support) but does not a PR to be merged (for now); those exceptional cases need a GitHub comment explaining them, of course.


> Most of the time it is one of us in the core team doing the review

Yes, I have set M-cleared-for-merge label after a review.

> > In other cases, a PR with a positive vote (implicit or explicit) and without negative ones, may have an M-cleared-for-merge label while it is waiting for various timeouts to expire (or other significant PR events).
 
> That is abusing Anubis and should ideally not be happening.

Why?! I do not see any signs of abuse here.


> Reviewers should at least wait until the automated tests have passed before spending time on review. 

I disagree that they should wait (in most cases), but if somebody wants to wait, they certainly can. There are no rules against that!


> A portability failure may cause major changes to the logic/design and re-reviewing such things is becoming a noticeable waste of our time.

Agreed. Each reviewer has to balance various risks/costs/factors. I see no relevance of any of this to M-cleared-for-merge. That label is not about review timing.


> > If you do not want to merge your own PR until somebody else (anybody) reviews it, do not clear it for merge, add a S-waiting-for-more-reviewers label, and a comment explaining this (rare) situation. That is what that label is for -- attract other reviewers to a PR that (effectively) does not have enough votes to be merged.

> Rule of thumb is that authors should not be clearing their own PRs, reviewers should be doing that. Though **after** the 10-day criteria passes it is reasonable to see that happen.

Labeling a PR with M-cleared-for-merge does not preclude others from reviewing that PR! That PR will wait at least 10 days. I do not understand why you want me to manually track which PRs have passed that 10-day criteria (so that I can label them then). Doing so will waste my time and delay merging of eligible PRs, with no positives to offset that harm (that cannot be accomplished by other/better means without these negative side effects).

If changing the spelling of M-cleared-for-merge would help, please propose better alternatives.

File: src/cf.data.pre
Comment (rousskov) [reply src/cf.data.pre]: > FWIW; I believe it was clearly implied by the old text stating the value as covering time from start of I/O with peer A to end of I/O with peer B.

This text is _not_ about switching from peer A to peer B (i.e. retrying peers using one FwdState instance). It is about re-forwarding requests using two FwdState instances. In some cases, Squid will use peer A, then peer B, and then peer A again, and peer B again. The original description did not document what would happen in those cases. An admin could guess correctly, but they had similar if not greater chances of guessing wrong.

> I do not think we need to go into the level of internal logic this 3rd paragraph has. All reasons for any re-forwarding, re-tries, and redirection are covered by and/or already alluded to by the new paragraph 2 statement "(including any retries)"

I think we do need to detail that because paragraph 2 only documents that the timer starts at A and stops at B. That documentation is not wrong (for some definition of a "retry"[^1]), but it does not cover what happens when there are two pairs of A and B events (A1 and B1 plus A2 and B2). Interpreting those two pairs as A1 and B2 would be wrong -- the timer is _not_ running between B1 and A2.

[^1]: Related Squid terminology is inconsistent, but we tend to call attempts within a single FwdState instance "retries" (e.g., connect_retries and retry_on_error).

If you would like, we can replace "starts" and "stops" in paragraph 2 with "starts (or resumes)" and "pauses", but I am not sure it will improve things -- I think it is best to keep that initial paragraph simple and document special cases later. And even if we do make those replacements in paragraph 2, I would still keep this paragraph (in some shape of form) to explain what is going on in those special (but not so rare in collapsed forwarding environments) re-forwarding cases. Said that, if you insist, I will remove this paragraph. Your call.

File: src/format/Format.cc
Comment (rousskov) [reply src/format/Format.cc:641]: > Do you have some cases where the `al->heir` is set without `al->request` being set?

I am not aware of any.

> AFAIK we should not have hierarchy contact without a request message, and the ALE should be setup with that. Perhapse it is possible to improve this code by just accessing the `al->request->heir` record and treating `al->heir` as deprecated ?

I hope there are better ways to address this problem than "treating ALE::hier as deprecated" (which, 
AFAICT, means _ignoring_ ALE::hier when logging `%<tt`).

The corresponding official code uses ALE::hier instead of ALE::request. Doing the opposite without serious analysis/tests/etc. is risky (and this PR does not take that risk). ALE::hier is used in several code contexts, including contexts unrelated to peering time. Ignoring it here may introduce bugs and/or more inconsistencies. Let this problem be handled in a dedicated PR (which will focus on addressing these issues properly rather than taking unnecessary risks).

File: src/format/Format.cc
Comment (rousskov) [reply src/format/Format.cc:647]: I agree that no conversion to timeval should be necessary, but improving Format::assemble() to handle std::chrono durations without such conversion should be done in a PR dedicated to that enhancement. This PR does not even introduce an std::chrono conversion to Format::assemble() -- it already exists in official Format::assemble() code.

File: src/cf.data.pre
Comment (rousskov) [reply src/cf.data.pre]: Done.

**Review by rousskov — COMMENTED** [review]
@yadij, I hope I have addressed all your change requests. Please re-review.

File: src/cf.data.pre
Comment (yadij) [reply src/cf.data.pre]: If that were true, you would not have insisted on me writing "server or peer" in all previous PRs.

Comment (rousskov) [reply src/cf.data.pre]: > > Alex: I have changed that brief `<tt` description to use the terminology you prefer: "Total time spent forwarding to origin servers or cache_peers (milliseconds)." I hope this addresses your concern.

> Amos: If that were true, you would not have insisted on me writing "server or peer" in all previous PRs.

I do not know what "it" refers to, but can we focus on moving forward this stalled PR rather than on my alleged past actions in other PRs, please? AFAICT, this change request has been addressed two months ago, using phrasing nearly identical to the one you have suggested. Please resolve this specific change request and, if possible, dismiss your negative PR review.

File: src/format/Format.cc
Comment (kinkie) [reply src/format/Format.cc:641]: I don't think this PR is the right place to deduplicate `al->request->heir` and `al->heir`.
There is a problem, and it should be tackled somewhere else (hopefully eventually)

File: src/format/Format.cc
Comment (kinkie) [reply src/format/Format.cc:647]: I'd resist scope creep here.
We are in the process of adding `std::chrono` and we are not consistent about it at all.
If @rousskov can, as he mentioned, add the facility to support it in a separate PR, I'm happy with that

File: src/cf.data.pre
Comment (kinkie) [reply src/cf.data.pre:5011]: Nobody likes the text, but it's a text and thus better than nothing.
I'll take a stab at improving it in a separate PR eventually, and we can discuss things there.

**Review by kinkie — APPROVED** [review]
While the documentation can be probably improved - noone really
likes it - it's not bad enough for me to reject the PR.
There will be followup PRs to do the followup, that's good enough for me

File: src/format/Format.cc
Comment (rousskov) [reply src/format/Format.cc:647]: > We will submit a dedicated followup PR offering a reusable version of this conversion

Done in #1915.

**Pull request closed without merging**
